apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/jvm/gc/GCModule.java

```diff
@@ -88,4 +98,12 @@ public List<GC> getGCList() {
     protected abstract String getNewGCName();
 
     protected abstract String getNormalGCName();
+
+    protected String getCycleGcBeanName() {
+        return null;
+    }
+
+    protected String getPauseGcBeanName() {
+        return null;
+    }
```

Comment:
I don't feel this is a good way. Couldn't you use an implementation to override `getNormalGCName`?

Reply:
the zgc gc count needs be collected from ZGC Cycles bean
the zgc gc pause time needs be collected from  ZGC Pauses  bean

so the two bean name should be in one gc module and then get  gc count and gc pause time from two bean  when collect

Reply:
I mean we should have zgc version detection. Right now, keeping adding new methods seems not a good code practice.

Reply:
add a new zgcmodule for jdk 17 and keep ZGC Cycles and  ZGC Pauses bean name in this module, when it is jdk17, set this module ？

Reply:
I don't have implementation preference as I don't know details enough.
Generally, don't add new methods for GCModule.

Reply:
gc time and count is calculated in GCModule.  now the GCModule method getOldGCName, getNewGCName, getNormalGCName is for get bean name and collect metirc from the target bean.

Generally zgc just need one bean(bean name from getNormalGCName method) to collect the gc time and gc count

but jdk17 zgc need two beans to calculate，i can not use other three exists method get the bean.

now i have two ways when  don't add new methods for GCModule

1. in GCModule, use string constant bean name not use method get it，i think it is strange

2. zgcmodule overwrite GCModule caluldate method and implementation itself. because zgc is  different with other gc. the GCModule  caluldate method is just universal for non zgc.

Reply:
My point is, can't you build another two new `ZGCModule` implementations in `GCProvider`? And those two return new values of `getNormalGCName`

Reply:
it can not, if you add two new ZGCModule，in current project, just one can be active，so in GCModule, getNormalGCName just get one bean ZGC Cycles or ZGC Pauses, but gc count is from ZGC Cycles and gc pause time is from ZGC Pauses, i think it is not working

Reply:
Then please override `public List<GC> getGCList` from `GCMetricAccessor` to create a new `ZGCModule`, then to build a new `List<GC>`. `lastNormalGCCount` and `lastNormalGCTime` could be simply removed as well.
That was why it was built in two layers of the abstract concept.